src/library/scala/collection/BitSetLike.scala

```diff
@@ -15,51 +15,57 @@ package collection
 
 import BitSetLike._
 import mutable.StringBuilder
+import scala.util.hashing.MurmurHash3
+import java.lang.{Long => jLong}
+import java.util
+
+import scala.runtime.AbstractFunction1
 
 /** A template trait for bitsets.
- *  $bitsetinfo
+ * $bitsetinfo
  *
  * This trait provides most of the operations of a `BitSet` independently of its representation.
  * It is inherited by all concrete implementations of bitsets.
  *
- *  @tparam  This the type of the bitset itself.
- *
- *  @define bitsetinfo
- *  Bitsets are sets of non-negative integers which are represented as
- *  variable-size arrays of bits packed into 64-bit words. The memory footprint of a bitset is
- *  determined by the largest number stored in it.
- *  @author  Martin Odersky
- *  @since 2.8
- *  @define coll bitset
- *  @define Coll `BitSet`
+ * @tparam  This the type of the bitset itself.
+ * @define bitsetinfo
+ * Bitsets are sets of non-negative integers which are represented as
+ * variable-size arrays of bits packed into 64-bit words. The memory footprint of a bitset is
+ * determined by the largest number stored in it.
+ * @author Martin Odersky
+ * @since 2.8
+ * @define coll bitset
+ * @define Coll `BitSet`
  */
-trait BitSetLike[+This <: BitSetLike[This] with SortedSet[Int]] extends SortedSetLike[Int, This] { self =>
+trait BitSetLike[+This <: BitSetLike[This] with SortedSet[Int]] extends SortedSetLike[Int, This] {
+  self =>
 
   def empty: This
+  protected[this] def copyOrSelf: This = fromBitMaskNoCopy(toBitMask)
 
   /** The number of words (each with 64 bits) making up the set */
   protected def nwords: Int
+  /** the index of highest word that has any bits set, or -1 if empty */
+  private[collection] def lastWordSet: Int = {
+    var idx = nwords - 1
+    while (idx >= 0 && word(idx) == 0L)
+      idx -= 1
+    idx
+  }
 
   /** The words at index `idx`, or 0L if outside the range of the set
-   *  '''Note:''' requires `idx >= 0`
+   * '''Note:''' requires `idx >= 0`
    */
   protected def word(idx: Int): Long
+  @inline private[collection] final def _word(idx: Int) = word(idx)
```

Comment:
What purpose does this method serve?

Reply:
It could be useful, performance wise, to have a `private[scala]` way to get the `elems` array out of the array backed implementations (`mutablle.BitSet` and `immutable.BitSetN`). Then, in places like  `BitSetLike.|`,  we could have a code path that loops over the array rather than making the virtual call to `def word` inside the loop.

Reply:
> What purpose does this method serve?

the word method is not visible to subclasses where referencing another `BitSet` 

> It could be useful, performance wise, to have a `private[scala]` way to get the `elems` array out of the array backed implementations (`mutablle.BitSet` and `immutable.BitSetN`). Then, in places like `BitSetLike.|`, we could have a code path that loops over the array rather than making the virtual call to `def word` inside the loop.

agreed, but it would block (or complicate) a `SparseBitSet` implementation, and mean that the three special cases would be moved into  `BitSetLike`. It doesn't seem that the abstractoin is great ATM though. `word` virtual dispatch is a bottleneck
It would be nice to restructure the hierarchy for the array backed implementations